src/python/pants/build_graph/build_graph.py

```diff
@@ -604,6 +605,17 @@ def inject_address_closure(self, address):
     """
 
   @abstractmethod
+  def inject_address_specs_closure(self, address_specs: Iterable[AddressSpec], fail_fast=None):
+    """Resolves, constructs and injects Targets and their transitive closures of dependencies.
+
+    :API: public
+
+    :param address_specs: An iterable of AddressSpec objects to resolve and inject.
+    :param fail_fast: Whether to fail quickly for the first error, or to complete all
+      possible injections before failing.
+    :returns: Yields a sequence of resolved Address objects.
+    """
+
   def inject_specs_closure(self, specs: Iterable[AddressSpec], fail_fast=None):
```

Comment:
I'm not sure how to deprecate this because it's an abstract method. If we keep it as `abstractmethod`, then any subclass _must_ implement the method, which we don't want.

I think the solution is to remove `abstractmethod` so that subclasses no longer need to define this. Where I'm unclear is if the below `warn_or_error()` will actually do anything because no subclass calls `super().inject_specs_closure`.

Reply:
> I think the solution is to remove abstractmethod so that subclasses no longer need to define this. Where I'm unclear is if the below warn_or_error() will actually do anything because no subclass calls super().inject_specs_closure.

Subclasses don't. But the reason it is marked `:API public:` is because Tasks outside of this repository might call it. So yea: can make the method concrete, and mark `@deprecated(self.FUTURE_VERSION, hint_message=hint_message)`.

Reply:
Thanks for explaining this one. Also TIL about `@deprecated` existing.

However, the implementation of `@deprecated` ends up being the same as what we have now using `warn_or_error` so I'm going to keep the current approach.

Thanks for the review!